src/lib/yacc/grammar.rs

```diff
@@ -699,6 +723,83 @@ fn nonterm_costs<F>(grm: &YaccGrammar, term_cost: F) -> Vec<u64>
     costs
 }
 
+
+/// Return the cost of the maximal string for each non-terminal in this grammar (u64::max_val()
+/// representing "this non-terminal can generate strings of infinite length"). The cost of a
+/// terminal is specified by the user-defined `term_cost` function.
+fn nonterm_max_costs(grm: &YaccGrammar, term_costs: &Vec<u64>) -> Vec<u64>
+{
+    let mut done = vec![];
+    done.resize(grm.nonterms_len(), false);
+    let mut costs = vec![];
+    costs.resize(grm.nonterms_len(), 0);
+
+    // First mark all recursive non-terminals.
+    for i in 0..grm.nonterms_len() {
+        // Calling has_path so frequently is not exactly efficient...
+        if grm.has_path(NTIdx::from(i), NTIdx::from(i)) {
+            costs[i] = u64::max_value();
+            done[i] = true;
+        }
+    }
+
+    loop {
+        let mut all_done = true;
+        for i in 0..done.len() {
+            if done[i] {
+                continue;
+            }
+            all_done = false;
+            let mut hs_cmplt = None; // highest completed cost
+            let mut hs_noncmplt = None; // highest non-completed cost
+            'a: for p_idx in grm.nonterm_to_prods(NTIdx::from(i)).iter() {
```

Comment:
Personally a fan of `'outer`.

Reply:
?

Reply:
Sorry, for being unspecific. I personally prefer more descriptive loop labels, such as `outer` instead of using just `a`.

Reply:
Each to their own. Given that you can't jump to an inner loop, I will personally admit to not being convinced by that particular naming scheme ;)